lib/jxl/enc_external_image.cc

```diff
@@ -110,11 +92,84 @@ Status ConvertFromExternal(Span<const uint8_t> bytes, size_t xsize,
 
   return true;
 }
+
+Status ConvertFromExternalNoSizeCheck(const uint8_t* data, size_t xsize,
+                                      size_t ysize, size_t stride,
+                                      const ColorEncoding& c_current,
+                                      size_t color_channels,
+                                      size_t bits_per_sample,
+                                      JxlPixelFormat format, ThreadPool* pool,
+                                      ImageBundle* ib) {
+  bool has_alpha = format.num_channels == 2 || format.num_channels == 4;
+  if (format.num_channels < color_channels) {
+    return JXL_FAILURE("Expected %" PRIuS
+                       " color channels, received only %u channels",
+                       color_channels, format.num_channels);
+  }
+
+  Image3F color(xsize, ysize);
+  for (size_t c = 0; c < color_channels; ++c) {
+    JXL_RETURN_IF_ERROR(ConvertFromExternalNoSizeCheck(
+        data, xsize, ysize, stride, bits_per_sample, format, c, pool,
+        &color.Plane(c)));
+  }
+  if (color_channels == 1) {
+    CopyImageTo(color.Plane(0), &color.Plane(1));
+    CopyImageTo(color.Plane(0), &color.Plane(2));
+  }
+  ib->SetFromImage(std::move(color), c_current);
+
+  // Passing an interleaved image with an alpha channel to an image that doesn't
+  // have alpha channel just discards the passed alpha channel.
+  if (has_alpha && ib->HasAlpha()) {
+    ImageF alpha(xsize, ysize);
+    JXL_RETURN_IF_ERROR(ConvertFromExternalNoSizeCheck(
+        data, xsize, ysize, stride, bits_per_sample, format,
+        format.num_channels - 1, pool, &alpha));
+    ib->SetAlpha(std::move(alpha));
+  } else if (!has_alpha && ib->HasAlpha()) {
+    // if alpha is not passed, but it is expected, then assume
+    // it is all-opaque
+    ImageF alpha(xsize, ysize);
+    FillImage(1.0f, &alpha);
+    ib->SetAlpha(std::move(alpha));
+  }
+
+  return true;
+}
+
+Status ConvertFromExternal(const uint8_t* data, size_t size, size_t xsize,
+                           size_t ysize, size_t bits_per_sample,
+                           JxlPixelFormat format, size_t c, ThreadPool* pool,
+                           ImageF* channel) {
+  size_t bytes_per_channel = JxlDataTypeBytes(format.data_type);
+  size_t bytes_per_pixel = format.num_channels * bytes_per_channel;
+  const size_t last_row_size = xsize * bytes_per_pixel;
+  const size_t align = format.align;
+  const size_t row_size =
+      (align > 1 ? jxl::DivCeil(last_row_size, align) * align : last_row_size);
```

Comment:
Does this PR happen to also fix #2971?

Reply:
Hopefully yes, I made the "onegroup" mode in the streaming test use >row size align value to test this.